core/src/callback.jl

```diff
@@ -141,6 +193,99 @@ function update_cumulative_flows!(u, t, integrator)::Nothing
+                    basin.mass[from_flow.idx, :] .-= placeholder .* flow
+                    @warn "Unsupported outflow type $(to_flow.type) #$(to_flow.value) with flow $flow"
+                end
+            end
+        end
+
+        if to_flow.type == NodeType.Basin
+            flow = flow_update_on_edge(integrator, outflow_edge.edge)
+            if flow > 0
+                basin.cumulative_in[to_flow.idx] += flow
+                if from_flow.type == NodeType.Basin
+                    basin.mass[to_flow.idx, :] .+=
+                        basin.concentration_state[from_flow.idx, :] .* flow
+                elseif from_flow.type == NodeType.LevelBoundary
+                    basin.mass[to_flow.idx, :] .+=
+                        level_boundary.concentration[from_flow.idx, :] .* flow
+                elseif from_flow.type == NodeType.UserDemand
+                    basin.mass[to_flow.idx, :] .+=
+                        user_demand.concentration[from_flow.idx, :] .* flow
+                elseif from_flow.type == NodeType.Terminal
+                    # TODO Apparently UserDemand outflow is discoupled from it's inflow
+                    # and its inflow_edge is Terminal #0 twice
+                    basin.mass[to_flow.idx, :] .+=
+                        user_demand.concentration[outflow_edge.edge[1].idx, :] .* flow
+                else
+                    @warn "Unsupported inflow type $(from_flow.type)"
+                end
+            end
+        end
+    end
+
+    basin.concentration_state .= basin.mass ./ (basin.storage_prev .+ basin.cumulative_in)
+
+    for (inflow_edge, outflow_edge) in zip(state_inflow_edge, state_outflow_edge)
+        from_flow = inflow_edge.edge[1]
+        to_flow = outflow_edge.edge[2]
+        if from_flow.type == NodeType.Basin
+            flow = flow_update_on_edge(integrator, inflow_edge.edge)
+            if flow > 0
+                basin.mass[from_flow.idx, :] .-=
+                    basin.concentration_state[from_flow.idx, :] .* flow
+            end
+        end
+
+        if to_flow.type == NodeType.Basin
+            flow = flow_update_on_edge(integrator, outflow_edge.edge)
+            if flow < 0
+                basin.mass[to_flow.idx, :] .+=
+                    basin.concentration_state[to_flow.idx, :] .* flow
+            end
+        end
+    end
+
+    if basin.evaporate_mass
+        basin.mass .-= basin.concentration_state .* (u.evaporation - uprev.evaporation)
+    end
+    basin.mass .-= basin.concentration_state .* (u.infiltration - uprev.infiltration)
+
+    # Take care of masses getting ever smaller, possibly becoming negative
+    # TODO Should this be bounded by the solver tolerances?
+    for I in eachindex(basin.mass)
+        if (0 - eps(Float64)) < basin.mass[I] < (0 + eps(Float64))
+            basin.mass[I] = 0.0
+        end
+    end
+    any(<(0), basin.mass) && error("Negative mass detected: $(basin.mass)")
```

Comment:
Although this error is probably rare, when it occurs it is not very informative in the current formulation. I would like to know when, in which basin which substance gets a negative mass.

Reply:
Agreed, it will now error out like this:

┌ Info: Starting a Ribasim simulation.
│   toml_path = "/Users/evetion/Downloads/hws_2024_7_x_delwaq/hws_24h.toml"
│   cli.ribasim_version = "2024.11.0"
│   starttime = 2023-01-20T00:00:00
└   endtime = 2023-07-01T00:00:00
┌ Error: Basin #670 has negative mass -1.0 for substance Initial
└ @ Ribasim /Users/evetion/code/Ribasim/core/src/callback.jl:316
┌ Error: Basin #1041 has negative mass -1.0 for substance UserDemand
└ @ Ribasim /Users/evetion/code/Ribasim/core/src/callback.jl:316
ERROR: Negative mass(es) detected